src/cachingreader.cpp

```diff
@@ -258,199 +266,161 @@ void CachingReader::process() {
+                    continue;
+                }
+                DEBUG_ASSERT(pChunk->state == Chunk::READ_IN_PROGRESS);
+                freeChunk(pChunk);
             }
-            DEBUG_ASSERT(pChunk->state == Chunk::READ_IN_PROGRESS);
-            freeChunk(pChunk);
         }
     }
 }
 
-int CachingReader::read(int sample, int num_samples, CSAMPLE* buffer) {
+int CachingReader::read(int sample, int numSamples, CSAMPLE* buffer) {
     // Check for bad inputs
-    DEBUG_ASSERT_AND_HANDLE(sample % 2 == 0) {
+    DEBUG_ASSERT_AND_HANDLE(sample % CachingReaderWorker::kChunkChannels == 0) {
         // This problem is easy to fix, but this type of call should be
         // complained about loudly.
         --sample;
     }
-    DEBUG_ASSERT_AND_HANDLE(num_samples % 2 == 0) {
-        --num_samples;
+    DEBUG_ASSERT_AND_HANDLE(numSamples % CachingReaderWorker::kChunkChannels == 0) {
+        --numSamples;
     }
-    if (num_samples < 0 || !buffer) {
+    if (numSamples < 0 || !buffer) {
         QString temp = QString("Sample = %1").arg(sample);
         qDebug() << "CachingReader::read() invalid arguments sample:" << sample
-                 << "num_samples:" << num_samples << "buffer:" << buffer;
+                 << "numSamples:" << numSamples << "buffer:" << buffer;
         return 0;
     }
 
     // If asked to read 0 samples, don't do anything. (this is a perfectly
     // reasonable request that happens sometimes. If no track is loaded, don't
     // do anything.
-    if (num_samples == 0 || m_readerStatus != TRACK_LOADED) {
+    if (numSamples == 0 || m_readerStatus != TRACK_LOADED) {
         return 0;
     }
 
     // Process messages from the reader thread.
     process();
 
-    int samples_remaining = num_samples;
+    SINT samplesRead = 0;
 
+    SINT frameIndex = samples2frames(sample);
+    SINT numFrames = samples2frames(numSamples);
 
     // TODO: is it possible to move this code out of caching reader
     // and into enginebuffer?  It doesn't quite make sense here, although
     // it makes preroll completely transparent to the rest of the code
     // if we're in preroll...
-    if (sample < 0) {
-        int zero_samples = math_min(-sample, samples_remaining);
-        DEBUG_ASSERT(0 <= zero_samples);
-        DEBUG_ASSERT(zero_samples <= samples_remaining);
-        SampleUtil::clear(buffer, zero_samples);
-        samples_remaining -= zero_samples;
-        if (samples_remaining == 0) {
-            //everything is zeros, easy
-            return zero_samples;
+    if (Mixxx::AudioSource::getMinFrameIndex() > frameIndex) {
+        const SINT prerollFrames = math_min(numFrames,
+                Mixxx::AudioSource::getMinFrameIndex() - frameIndex);
+        const SINT prerollSamples = frames2samples(prerollFrames);
+        SampleUtil::clear(buffer, prerollSamples);
+        if (prerollSamples == numSamples) {
+            return numSamples; // early exit
```

Comment:
The comment explains the keyword return. ;-) What about: 
// All requests samples are in the pre-roll: done!